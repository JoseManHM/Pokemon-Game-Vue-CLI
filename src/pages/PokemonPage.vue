<template>
    <h1 v-if="!pokemon">Cargando...</h1>
    <div v-else>
        <h1>¿Quién es este pokemon??</h1>
        <!-- TODO: img -->
        <pokemon-image 
            :pokemonId="pokemon.id" :showPokemon="showPokemon" 
        />
        <!--TODO: opciones -->
        <pokemon-options 
            :pokemons="pokemonArr"
            @selection="checkAnswer"
        />
        <template v-if="showAnswer">
            <h2 class="fade-in">{{message}}</h2>
            <button @click="newGame">Nuevo juego</button>
        </template>
    </div>
</template>

<script>
import PokemonImage from '@/components/PokecmonImage'
import PokemonOptions from '@/components/PokemonOptions'
import getPokemonOptions from '@/helpers/getPokemonOptions'

//console.log( getPokemonOptions() )

export default {
    components:{PokemonImage, PokemonOptions},
    data(){
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: '',
            lives: 1
        }
    },
    methods: {
        async mixPokemonArray(){
            this.pokemonArr = await getPokemonOptions()
            
            const randomNumber = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[randomNumber]
        },
        checkAnswer: function(selectedID){
            if(selectedID == this.pokemon.id){
                this.showPokemon = true
                this.showAnswer = true
                this.message = `Correcto, ${this.pokemon.name}`
            }else{
                if(this.lives == 0){
                    this.showPokemon = true
                    this.showAnswer = true
                    this.message = `Ups, la opción correcta es ${this.pokemon.name}`
                }else{
                    this.showPokemon = false
                    this.showAnswer = true
                    this.message = `Ups, te queda una vida`
                    this.lives --
                }
            }
        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr =[]
            this.pokemon = null
            this.lives = 1
            this.mixPokemonArray()
        }
    },
    mounted(){
        this.mixPokemonArray()
    }
}
</script>