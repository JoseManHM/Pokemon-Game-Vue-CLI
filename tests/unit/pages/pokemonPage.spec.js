import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/pokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

import PokemonImage from '@/components/PokecmonImage'
import PokemonOptions from '@/components/PokemonOptions'

describe('Pokemon Page component', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('Debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de llamar el mixPokemonArray al montar', () => {
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('Debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    lives: 1
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    lives: 1
                }
            }
        })
        let componentImage = wrapper.findComponent(PokemonImage)
        let componentOptions = wrapper.findComponent(PokemonOptions)
        //PokemonPicture debe de existir
        expect(componentImage.exists()).toBe(true)
        //PokemonOptions debe de existir
        expect(componentOptions.exists()).toBe(true)
        
        //PokemonPicture atributo pokemonId === 5
        expect(componentImage.attributes('pokemonid')).toBe("5")
        //PokemonOptions atributo pokemons toBe true
        expect(componentOptions.attributes('pokemons')).toBeTruthy()

        //Otra forma de resolver la prueba
        let picture = wrapper.find('pokemon-image-stub')
        let options = wrapper.find('pokemon-options-stub')
        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()
        //Propiedades
        expect(picture.attributes('pokemonid')).toBe("5")
        expect(options.attributes('pokemons')).toBeTruthy()
    })

    test('Pruebas con checkAnswer', async() => {
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    lives: 1
                }
            }
        })
        //Cuando la respuesta es correcta
        await wrapper.vm.checkAnswer(5)
        expect(wrapper.vm.pokemon.id).toBe(pokemons[0].id)
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.vm.showAnswer).toBeTruthy()
        expect(wrapper.vm.message).toBe(`Correcto, ${pokemons[0].name}`)
        expect(wrapper.vm.lives).toBe(1)
        //Cuando la respuesta es incorrecta
        await wrapper.vm.checkAnswer(10)
        expect(wrapper.vm.pokemon.id).toBe(pokemons[0].id)
        expect(wrapper.vm.showPokemon).toBe(false)
        expect(wrapper.vm.showAnswer).toBe(true)
        expect(wrapper.vm.message).toBe('Ups, te queda una vida')
        expect(wrapper.vm.lives).toBe(0)
    })
})