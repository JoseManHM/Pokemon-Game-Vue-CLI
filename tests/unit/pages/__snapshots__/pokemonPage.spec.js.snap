// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Page component Debe de hacer match con el snapshot 1`] = `<h1>Cargando...</h1>`;

exports[`Pokemon Page component Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokemon??</h1><!-- TODO: img -->
  <pokemon-image-stub pokemonid="5" showpokemon="false"></pokemon-image-stub>
  <!--TODO: opciones -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
